### PR TITLE
RegularGridTopology dimensions

### PR DESCRIPTION
**Pull request by epernod**
Small PR on GridTopology and RegularGridTopology link to issue: #163

- Allow to set RegularGridTopology in 1,2 or 3 dimension. But still check if the grid resolution is valid.
- Fix example of Tetrahedron FEM that was crashing if no tetrahedron is found in the scene.
- Add an example scene.
- Remove RegularGridTopology dependencies from MechanicalObject.

CHANGELOG:
  - GridTopology 
       - clean the code & factor the constructor
       - implement "flat" grids in 1 or 2 dimension by using setting grid resolution to "1" in the corresponding axis
  - RegularGridTopology clean the constructor's code & remove NDEBUG code. 
  - Adds RegularGridTopology_test
  - TetrahedronFEMForceField clean code
  - MechanicalObject removes code that was specific to the grid. 
  - Adds new examples scenes called: 
         examples/Components/topology/Hexa2TetraTopologicalMapping.scn
         examples/Components/topology/RegularGridTopology_dimension.scn
______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings nor unit test failures.
- [x] does not break existing scenes.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

File: SofaKernel/modules/SofaBaseTopology/GridTopology.cpp
Comment (damienmarchal) [inline SofaKernel/modules/SofaBaseTopology/GridTopology.cpp]: If the object inherit from Base you can now write 
msg_warning() instead of msg_warning(this)

File: applications/projects/SofaPhysicsAPI/SofaPhysicsSimulation.cpp
Comment (damienmarchal) [inline applications/projects/SofaPhysicsAPI/SofaPhysicsSimulation.cpp]: This looks like debugging so it should probably be removed or replaced with msg_info(...)

**Review by damienmarchal — COMMENTED** [review]
Thank @epernod  for PR, it fix existing problems, add examples and is good work. 
If you have time you can add a GridTopology_test somewhere to avoid regression and if you don't I will probably do one someday :)

File: applications/projects/SofaPhysicsAPI/SofaPhysicsSimulation.cpp
Comment (epernod) [reply applications/projects/SofaPhysicsAPI/SofaPhysicsSimulation.cpp]: totally debug... didn't see that commit end up in this PR ... Will remove that

**Comment by epernod** [comment]
@damienmarchal I have added a RegularGridTopology_test. In fact I don't think GridTopology is used like this as a component. It is more a mother class for Regular/Cylindrical and sphereGridTopology (which each has tests now)

**Comment by damienmarchal** [comment]
Thanks Eric for the fixes and the additional test.

Your answer about GridTopology rise other questions:
   - If the component is not supposed to be used without being inherited... shouldn't we remove it from the factory  ?
   - If a class is not a component does not mean we shouldn't test it. Actually my feeling is that base class (as they are shared) are very good entities to test :) 

EDIT: Maybe these questions should be for a new Issue and not to handle in this PR.

**Comment by epernod** [comment]
In fact most of the methods I'm testing are directly inside the mother class. So I'm not sure a gridTopology_test is needed.
But to remove GridTopology from the factory is maybe possible. We need to investigate if someone is using it as a component.

**Comment by damienmarchal** [comment]
Hi Erik, 

There is one test crash in  TetrahedronFEMForceField_test. Is this related to your changes ?

DM

**Comment by epernod** [comment]
Should not but I will check.
Yes in fact my fix in TetrahedronFEMFF reveals that the test was doing nothing: This the log on the master branch:
```
[ RUN      ] TetrahedronDiffusionFEMForceField_test/0.extension
[ERROR]   [TetrahedronFEMForceField(fem)]  object must have a mesh topology. The component is inactivated.  To remove this error message please add a topology component to your scene.

```
Right now it crashes... will see if I understand enough the test to fix it.

**Comment by guparan** [comment]
Hi @epernod 
Is this PR also linked to #162 ?

**Comment by damienmarchal** [comment]
Hi @epernod 

Can I add more tests on this PR ? Because I think we should validate a bit more "negative/bad" case  to enforce the behavior of component when used improperly.

**Comment by epernod** [comment]
help yourself.

File: SofaKernel/modules/SofaBaseTopology/GridTopology.cpp
Comment (matthieu-nesme) [inline SofaKernel/modules/SofaBaseTopology/GridTopology.cpp]: prefer beginWriteOnly vs beginEdit when possible it saves an update.

File: SofaKernel/modules/SofaBaseTopology/GridTopology.h
Comment (matthieu-nesme) [inline SofaKernel/modules/SofaBaseTopology/GridTopology.h]: what is the interest of storing this information?

**Review by matthieu-nesme — DISMISSED** [review]

File: SofaKernel/modules/SofaBaseTopology/GridTopology.cpp
Comment (epernod) [reply SofaKernel/modules/SofaBaseTopology/GridTopology.cpp]: ok. Done.

File: SofaKernel/modules/SofaBaseTopology/GridTopology.h
Comment (epernod) [reply SofaKernel/modules/SofaBaseTopology/GridTopology.h]: First idea was to use it from other components, like a tetrahedronFEM check that the grid is 3D to avoid crash.
But while testing I saw that we directly do test on the topology buffers.
So right now it is not used excepts in the test @damienmarchal added. But that the idea.

Comment (damienmarchal) [reply SofaKernel/modules/SofaBaseTopology/GridTopology.h]: Up to my understanding, GridTopology have dimension once they are initialized...this is property of the topology that can be  recomputed from the nx,ny,nz each time we need it or stored. I have not opinion on that as long as there is function to retrieve the dimension.

**Comment by damienmarchal** [comment]
@matthieu-nesme  
Do you think we should remove the dimension ? 
Or is it ok to have this PR switched to the ready state ?

**Comment by matthieu-nesme** [comment]
To me the notion of dimension is not well defined, keeping it as a function w/o Data would be better.
`unsigned getMaterialDimension() const` ?

**Comment by untereiner** [comment]
I would suggest `unsigned getDimension() const` because the object should be considered as a topological space of a given dimension and not as a material since it does not have material properties yet.

**Comment by epernod** [comment]
I agree, "Material" does not mean something in this base component. 
This parameter should only be informative, not a Data, there is no mechanism to change the grid resolution if you change this dimension.

**Comment by epernod** [comment]
Finally do we have an agreement here? @matthieu-nesme @untereiner @damienmarchal 
Or do we remove that field?

**Comment by damienmarchal** [comment]
getDimmension() sounds ok to me. 
Not needed to have it as a Data. 
It can be recomputed each time getDimmension is called, there is no need to store it.

**Comment by damienmarchal** [comment]
@epernod, @matthieu-nesme  do you agreed so that we can merge soon ?

**Comment by epernod** [comment]
It is ok for me. We are just speaking about an informative int or an enum. 
I can do the change (if any) as soon as their is a final decision.

**Comment by damienmarchal** [comment]
@hugtalbot, you set this PR to ready but didn't dismissed @matthieu-nesme's "request for change", it  looks a bit ambiguous to me. 

Let's take it easy, I will dismissed the request for change and do the merge. If this is not what was intented don't hesitate to revert or submit a small fix to remove the Grid_dimmension as matthieu suggested.

**Comment by damienmarchal** [comment]
[ci-build]

**Comment by matthieu-nesme** [comment]
I still do not get the interest of storing `m_gridDim`. What is the limitation of a simple function? So you never have to change the value of this variable to keep it consistent.

If it was in a data you want to plug in a Data/Engine graph, it would indeed be necessary to store it, but it is not the case here.

**Comment by damienmarchal** [comment]
Everyone agreed there is no interest in storing m_gridDim.  
9 days ago erik asked:
" Or do we remove that field?"  followed by "I can do the change (if any) as soon as their is a final decision."

He is waiting for an answer since 9 days.  
So here is the answer, "yes do the change and remove this field" :)

**Comment by epernod** [comment]
yes we can change the test to remove` badDim` as now the dim is not computed at the same time.

No, `_n[i] = 1` works. `n = [1; 5; 5]` give a plan grid of 25 cells. -> Dim = 2

**Comment by matthieu-nesme** [comment]
@epernod If I am not wrong, if you have n=[2,2,2] your code will give dim=0 where it should give dim=3.
Only a test could check that ;)

**Comment by epernod** [comment]
Indeed... my bad.
This Pr already add tests for the gridTopology but the Tests needed a test. 
Thanks @matthieu-nesme for the review. It should be good now.

**Comment by damienmarchal** [comment]
[ci-build] 

EDIT: Thank you all to make things moving and progressing.

**Comment by damienmarchal** [comment]
Hi Erik,

It seems that this PR introduced a new examples RegularGrid_dimmension that is failing because it use a component ```<EdgeSet/>``` that does not seems to be part of Sofa.

Can you provide a bit of details please.

**Pull request merged**
